src/main/java/com/google/devtools/build/lib/analysis/actions/StrippingPathMapper.java

```diff
@@ -0,0 +1,268 @@
+ * </ol>
+ *
+ * <p>So an action is responsible for declaring that it strips paths and adjusting its command line
+ * accordingly. The executor is responsible for remapping action inputs and outputs to match.
+ */
+public final class StrippingPathMapper {
+  static final String GUID = "8eb2ad5a-85d4-435b-858f-5c192e91997d";
+
+  /**
+   * Creates a new {@link PathMapper} that strips config prefixes if the particular action instance
+   * supports it.
+   *
+   * @param action the action to potentially strip paths from
+   * @return a {@link StrippingPathMapper} if the action supports it, else {@link Optional#empty()}.
+   */
+  static Optional<PathMapper> tryCreate(Action action) {
+    // This is expected to always be "bazel-out", but we don't want to hardcode it here.
+    PathFragment outputRoot = action.getPrimaryOutput().getExecPath().subFragment(0, 1);
+    if (isPathStrippable(action.getInputs(), outputRoot)) {
+      return Optional.of(
+          create(action.getMnemonic(), action instanceof StarlarkAction, outputRoot));
+    }
+    return Optional.empty();
+  }
+
+  private static PathMapper create(String mnemonic, boolean isStarlarkAction,
+      PathFragment outputRoot) {
+    final StringStripper argStripper = new StringStripper(outputRoot.getPathString());
+    final ExceptionlessMapFn<Object> structuredArgStripper =
+        (object, args) -> {
+          if (object instanceof String) {
+            args.accept(argStripper.strip((String) object));
+          } else {
+            args.accept(CommandLineItem.expandToCommandLine(object));
+          }
+        };
+    // This kind of special handling should not be extended. It is a hack that works around a
+    // limitation of the native implementation of location expansion: The output is just a list of
+    // strings, not a structured command line that would allow transparent path mapping.
+    // Instead, reimplement location expansion in Starlark and have it return an Args object.
+    final boolean isJavaAction =
+        mnemonic.equals("Javac") || mnemonic.equals("JavacTurbine") || mnemonic.equals("Turbine");
+    return new PathMapper() {
+      @Override
+      public String getMappedExecPathString(ActionInput artifact) {
+        if (artifact instanceof DerivedArtifact || artifact instanceof ParamFileActionInput) {
+          return strip(artifact.getExecPath()).getPathString();
+        } else {
+          return artifact.getExecPathString();
+        }
+      }
+
+      @Override
+      public PathFragment map(PathFragment execPath) {
+        return isOutputPath(execPath, outputRoot) ? strip(execPath) : execPath;
+      }
+
+      @Override
+      public List<String> mapCustomStarlarkArgs(List<String> args) {
+        if (!isStarlarkAction) {
+          return args;
+        }
+        // Add your favorite Starlark mnemonic that needs custom arg processing here.
+        if (!mnemonic.contains("Android")
```

Comment:
adding more mmemonics + startlarkArgsToStrip here:
* mnemonics: `StarlarkAARGenerator`, `JetifySrcs` & `Desugar`
* starlarkArgsToStrip = `--mainData` ,`-i`, `--input`

Reply:
Sounds good to me, thanks!